build-tools/packages/build-cli/src/commands/generate/typetests.ts

```diff
@@ -0,0 +1,189 @@
+			level,
+		);
+		this.verbose(`Found ${previousPackageLevel} type definitions at: ${previousTypeDefs}`);
+
+		const { currentFile, previousFile } = initializeProjectsAndLoadFiles(
+			previousTypeDefs,
+			previousBasePath,
+			this.logger,
+		);
+
+		this.verbose(`Loaded source file for current version: ${currentFile.getFilePath()}`);
+		this.verbose(`Loaded source file for previous version: ${previousFile.getFilePath()}`);
+
+		const currentTypeMap = typeDataFromFile(currentFile);
+		const previousData = [...typeDataFromFile(previousFile).values()];
+
+		// eslint-disable-next-line unicorn/consistent-function-scoping
+		function compareString(a: string, b: string): number {
+			return a > b ? 1 : a < b ? -1 : 0;
+		}
+		previousData.sort((a, b) => compareString(a.name, b.name));
+
+		const fileHeader: string[] = [
+			`
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/*
+ * THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+ * Generated by fluid-type-test-generator in @fluidframework/build-tools.
+ */
+
+import type * as old from "${previousPackageName}${
+				previousPackageLevel === ApiLevel.public ? "" : `/${previousPackageLevel}`
+			}";
+import type * as current from "../../index.js";
+		`.trim(),
+			"\n",
+			typeOnly,
+		];
+
+		const testCases = generateCompatibilityTestCases(
+			previousData,
+			currentTypeMap,
+			currentPackageJson,
+			fileHeader,
+		);
+
+		mkdirSync("./src/test/types", { recursive: true });
+
+		writeFileSync(typeTestOutputFile, testCases.join("\n"));
+		console.log(`generated ${path.resolve(typeTestOutputFile)}`);
+	}
+
+	private getTypesPathWithFallback(
+		packageJson: PackageJson,
+		level: ApiLevel,
+	): [string, ApiLevel] {
+		let chosenLevel: ApiLevel = level;
+		// First try the requested paths, but fall back to public otherwise if configured.
+		let typeDefs: string | undefined = getTypesPathFromPackage(packageJson, level);
+
+		if (typeDefs === undefined) {
+			// Try the public types if configured to do so. If public types are found adjust the level accordingly.
+			typeDefs = this.flags.publicFallback
+				? getTypesPathFromPackage(packageJson, ApiLevel.public)
+				: undefined;
+			if (typeDefs === undefined) {
+				this.error(`No type definitions found for ${packageJson.name}`, { exit: 1 });
+			}
+			chosenLevel = ApiLevel.public;
+		}
+		return [typeDefs, chosenLevel];
+	}
+}
+
+/**
+ * Finds the path to the types of a package using the package's export map.
+ * If the path is found, it is returned. Otherwise, an error is thrown.
+ *
+ * This implementation uses resolve.exports to resolve the path to types for a level.
+ *
+ * @param packageJson - The package.json object to check for types paths.
+ * @param level - An API level to get types paths for.
+ * @returns A package relative path to the types.
+ */
+export function getTypesPathFromPackage(packageJson: PackageJson, level: ApiLevel): string {
```

Comment:
This function could also live in typeTestUtils. I think there is a more generalized version of this function that could be shared in several places, but it may make more sense for those callsites to just use resolve.exports directly.